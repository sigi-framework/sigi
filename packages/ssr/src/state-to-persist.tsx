import { GLOBAL_KEY_SYMBOL, RETRY_KEY_SYMBOL } from '@sigi/core'
import serialize from 'serialize-javascript'

const ScriptId = 'sigi-persisted-data'

export class StateToPersist<T> {
  constructor(
    private readonly dataToPersist: T,
    private readonly actionsToRetry: { [index: string]: string[] },
  ) {}

  extractToScriptString(withScriptTag = true) {
    if (this.dataToPersist == null) {
      return ''
    }
    const contentScript = this.serialize()
    return withScriptTag ? `<script id="${ScriptId}">${contentScript}</script>` : contentScript
  }

  renderToJSX() {
    if (this.dataToPersist == null) {
      return null
    }
    return <script id={ScriptId} dangerouslySetInnerHTML={{ __html: this.serialize() }} />
  }

  renderToDocument(doc: string) {
    if (this.dataToPersist == null) {
      return doc
    }
    const endBodyPosition = doc.indexOf('</body>')
    if (endBodyPosition === -1) {
      return doc
    }
    return doc.substr(0, endBodyPosition) + this.extractToScriptString() + doc.substring(endBodyPosition)
  }

  extractToJSONScriptString() {
    if (this.dataToPersist == null) {
      return ''
    }
    const stateContent = serialize(this.dataToPersist, { isJSON: true })
    return `<script id="${GLOBAL_KEY_SYMBOL}" type="application/json">${stateContent}</script><script id="${RETRY_KEY_SYMBOL}" type="application/json">${JSON.stringify(
      this.actionsToRetry,
    )}</script>`
  }

  private serialize() {
    const content = serialize(this.dataToPersist, { isJSON: true })
    return `window['${GLOBAL_KEY_SYMBOL}']=${content};window['${RETRY_KEY_SYMBOL}']=${JSON.stringify(
      this.actionsToRetry,
    )}`
  }
}
