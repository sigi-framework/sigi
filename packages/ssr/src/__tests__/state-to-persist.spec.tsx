import { renderToString } from 'react-dom/server'

import { StateToPersist } from '../state-to-persist'

describe('StateToPersist specs', () => {
  function createStateToPersist<T>(data?: T, actionsToRetry = {}) {
    return new StateToPersist(data, actionsToRetry)
  }

  const renderDocumentJSX = (child: React.ReactNode | null = null) => (
    <html>
      <head>
        <link rel="stylesheet" href="https://sigi.how/style.css" />
      </head>
      <body>
        <div id="app">
          <img src="https://sigi.how/image.jpeg" alt="" />
        </div>
        {child}
      </body>
    </html>
  )

  it('empty data should return empty string', () => {
    const state = createStateToPersist()
    expect(state.extractToScriptString()).toBe('')
  })

  it('should be able to extractScriptString without tag', () => {
    const state = createStateToPersist({ bar: 'baz' })
    expect(state.extractToScriptString(false)).toMatchSnapshot()
  })

  it('empty data should return original string', () => {
    const state = createStateToPersist()
    const originalString = renderToString(renderDocumentJSX())
    expect(state.renderToDocument(originalString)).toBe(originalString)
  })

  it('should render to document', () => {
    const state = createStateToPersist({ foo: 1 })
    expect(state.renderToDocument(renderToString(renderDocumentJSX()))).toMatchSnapshot()
  })

  it('should return original string if contains non </body> tag', () => {
    const state = createStateToPersist({ foo: 1 })
    const jsx = (
      <div>
        <span>1</span>
      </div>
    )
    expect(state.renderToDocument(renderToString(jsx))).toBe(renderToString(jsx))
  })

  it('should prevent xss', () => {
    const state = createStateToPersist({
      xss: '</script><script>alert(1);</script>',
    })
    expect(state.renderToDocument(renderToString(renderDocumentJSX()))).toMatchSnapshot()
  })

  it('should return null in renderToJSX if data is null/undefined', () => {
    const state = createStateToPersist()
    expect(state.renderToJSX()).toBeNull()
  })

  it('should be able to renderToJSX', () => {
    const state = createStateToPersist({ foo: 1 })
    expect(renderToString(renderDocumentJSX(state.renderToJSX()))).toMatchSnapshot()
  })

  it('should be able extract actionsToRetry', () => {
    const state = createStateToPersist({ foo: 1 }, { home: ['initAppContext'] })
    expect(renderToString(renderDocumentJSX(state.renderToJSX()))).toMatchSnapshot()
  })

  it('empty data should return empty string', () => {
    const state = createStateToPersist()
    expect(state.extractToJSONScriptString()).toBe('')
  })

  it('should be able to extractToJSONScriptString', () => {
    const state = createStateToPersist({ foo: 'bar' })
    expect(state.extractToJSONScriptString()).toMatchSnapshot()
  })
})
