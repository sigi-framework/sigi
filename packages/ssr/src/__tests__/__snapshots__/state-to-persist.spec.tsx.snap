// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StateToPersist specs should be able extract actionsToRetry 1`] = `"<html><head><link rel="preload" as="image" href="https://sigi.how/image.jpeg"/><link rel="stylesheet" href="https://sigi.how/style.css"/></head><body><div id="app"><img src="https://sigi.how/image.jpeg" alt=""/></div><script id="sigi-persisted-data">window['SIGI_STATE']={"foo":1};window['SIGI_RETRY']={"home":["initAppContext"]}</script></body></html>"`;

exports[`StateToPersist specs should be able to extractScriptString without tag 1`] = `"window['SIGI_STATE']={"bar":"baz"};window['SIGI_RETRY']={}"`;

exports[`StateToPersist specs should be able to extractToJSONScriptString 1`] = `"<script id="SIGI_STATE" type="application/json">{"foo":"bar"}</script><script id="SIGI_RETRY" type="application/json">{}</script>"`;

exports[`StateToPersist specs should be able to renderToJSX 1`] = `"<html><head><link rel="preload" as="image" href="https://sigi.how/image.jpeg"/><link rel="stylesheet" href="https://sigi.how/style.css"/></head><body><div id="app"><img src="https://sigi.how/image.jpeg" alt=""/></div><script id="sigi-persisted-data">window['SIGI_STATE']={"foo":1};window['SIGI_RETRY']={}</script></body></html>"`;

exports[`StateToPersist specs should prevent xss 1`] = `"<html><head><link rel="preload" as="image" href="https://sigi.how/image.jpeg"/><link rel="stylesheet" href="https://sigi.how/style.css"/></head><body><div id="app"><img src="https://sigi.how/image.jpeg" alt=""/></div><script id="sigi-persisted-data">window['SIGI_STATE']={"xss":"\\u003C\\u002Fscript\\u003E\\u003Cscript\\u003Ealert(1);\\u003C\\u002Fscript\\u003E"};window['SIGI_RETRY']={}</script></body></html>"`;

exports[`StateToPersist specs should render to document 1`] = `"<html><head><link rel="preload" as="image" href="https://sigi.how/image.jpeg"/><link rel="stylesheet" href="https://sigi.how/style.css"/></head><body><div id="app"><img src="https://sigi.how/image.jpeg" alt=""/></div><script id="sigi-persisted-data">window['SIGI_STATE']={"foo":1};window['SIGI_RETRY']={}</script></body></html>"`;
